### ISSUE TITLE
`<variant>`: Error with non-trivial types

### ISSUE THREAD

**Issue by nico-engels** [StephanTLavavej/STL#19]
# Basic Info
* [x] I have read the rules for the [Standard Library Modules Bug Bash][bug-bash-url].

* [x] I have pulled and rebuilt the `bug-bash` branch in the last 24 hours.

* My current commit, as printed by `git rev-parse HEAD`, is: e7b01974b8835b033891d926ff8a29eec10bc714

* This bug is:
rejects-valid

# Command-Line Test Case
My test case that works with classic includes but fails with named modules:

```
D:\cpp\STL\cpp-meo>type bug-3.cpp
```
```cpp
#ifdef USE_CLASSIC_INCLUDES
#include <variant>
#else
import std;
#endif

struct A
{
  ~A() { };
  //~A() = default; // works
};

int main()
{
    [[maybe_unused]] std::variant<A> v;
}
```
```
D:\cpp\STL\cpp-meo>cl /EHsc /nologo /W4 /std:c++latest /reference ..\std.ifc /MTd /Od bug-3.cpp ..\std.obj /DUSE_CLASSIC_INCLUDES && bug-3.exe
bug-3.cpp

D:\cpp\STL\cpp-meo>cl /EHsc /nologo /W4 /std:c++latest /reference ..\std.ifc /MTd /Od bug-3.cpp ..\std.obj && bug-3.exe
bug-3.cpp
D:\cpp\STL\out\x64\out\inc\variant(844): error C3861: '_Destroy_in_place': identificador não encontrado
D:\cpp\STL\out\x64\out\inc\variant(590): note: consulte a referência à instanciação 'auto std::_Variant_base<A>::_Destroy::<lambda_1>::operator ()<std::_Tagged<_First &,0>>(_T1) noexcept const' do modelo que está sendo compilada
        with
        [
            _First=A,
            _T1=std::_Tagged<A &,0>
        ]
D:\cpp\STL\out\x64\out\inc\variant(708): note: consulte a referência à instanciação 'std::_Variant_raw_visit_t<std::_Variant_base<A>::_Destroy::<lambda_1>,std::_Variant_storage_<false,A>&>' do modelo da alias que está sendo compilada
D:\cpp\STL\out\x64\out\inc\variant(708): note: consulte a referência à instanciação 'unknown-type std::_Variant_raw_visit(unsigned __int64,_Storage &&,_Fn &&) noexcept(<expr>)' do modelo que está sendo compilada
D:\cpp\STL\out\x64\out\inc\variant(840): note: ao compilar a função de membro 'void std::_Variant_base<A>::_Destroy(void) noexcept' do modelo da classe
D:\cpp\STL\out\x64\out\inc\variant(897): note: consulte a referência à instanciação 'void std::_Variant_base<A>::_Destroy(void) noexcept' do modelo que está sendo compilada
D:\cpp\STL\out\x64\out\inc\variant(893): note: consulte a referência à instanciação 'std::_Variant_base<A>' do modelo da classe que está sendo compilada
D:\cpp\STL\out\x64\out\inc\xsmf_control.h(24): note: consulte a referência à instanciação 'std::_Variant_destroy_layer_<A>' do modelo da classe que está sendo compilada
D:\cpp\STL\out\x64\out\inc\xsmf_control.h(44): note: consulte a referência à instanciação 'std::_Non_trivial_copy<std::_Variant_destroy_layer_<A>>' do modelo da classe que está sendo compilada
D:\cpp\STL\out\x64\out\inc\xsmf_control.h(65): note: consulte a referência à instanciação 'std::_Non_trivial_move<std::_Variant_destroy_layer_<A>,A>' do modelo da classe que está sendo compilada
D:\cpp\STL\out\x64\out\inc\xsmf_control.h(103): note: consulte a referência à instanciação 'std::_Non_trivial_copy_assign<std::_Variant_destroy_layer_<A>,A>' do modelo da classe que está sendo compilada
D:\cpp\STL\out\x64\out\inc\variant(983): note: consulte a referência à instanciação 'std::_Non_trivial_move_assign<std::_Variant_destroy_layer_<A>,A>' do modelo da classe que está sendo compilada
bug-3.cpp(16): note: consulte a referência à instanciação 'std::variant<A>' do modelo da classe que está sendo compilada
D:\cpp\STL\out\x64\out\inc\variant(842): error C2672: '_Variant_raw_visit': nenhuma função sobrecarregada correspondente encontrada
D:\cpp\STL\out\x64\out\inc\variant(708): note: poderia ser 'unknown-type std::_Variant_raw_visit(unsigned __int64,_Storage &&,_Fn &&) noexcept(<expr>)'
```
# Additional Context
With types that don't trigger the non_trivial machinery it works. First identified the problem with `std::string`.

[bug-bash-url]: https://github.com/StephanTLavavej/STL/wiki/Standard-Library-Modules-Bug-Bash

**Comment by StephanTLavavej** [comment]
Thanks! :smile_cat: The root cause appears to be [VSO-1538668](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1538668) "Compiler does not bind names inside of lambda nested in function template with parameter pack", fixed for 17.4 Preview 1. This was originally seen and worked around in `<scoped_allocator>`. The `<variant>` code is slightly different:

https://github.com/StephanTLavavej/STL/blob/e7b01974b8835b033891d926ff8a29eec10bc714/stl/inc/scoped_allocator#L227-L236
https://github.com/StephanTLavavej/STL/blob/e7b01974b8835b033891d926ff8a29eec10bc714/stl/inc/variant#L840-L848

They both involve generic lambdas, but the `<scoped_allocator>` occurrence involves a member function that's a variadic template, while the `<variant>` occurrence is a non-templated member function of a class that's a variadic template. I believe that these are sufficiently similar as to be not worth a separate compiler bug report.

I'll look into pushing a library workaround soon.

**Comment by StephanTLavavej** [comment]
Pushed a workaround - your test case compiles cleanly now. Thanks again!

**Comment by StephanTLavavej** [comment]
I've verified that the compiler fix has shipped in 17.4 Preview 1 and I've removed the library workaround.
